src/agent/block_store_services/block_store_client.js

```diff
@@ -120,52 +119,51 @@ class BlockStoreClient {
-
-                        // if not OK parse the error and throw Error object
-                        if (res.statusCode !== 200) {
-                            await this._throw_s3_err(res);
-                        }
-
+                    if (!this.block_to_creds) {
+                        this.block_to_creds = {};
+                    }
+                    if (!this.block_to_creds[options.address]) {
+                        this.block_to_creds[options.address] = await rpc_client.block_store.get_block_store_info(null, options);
                     }
+                    const s3_params = this.block_to_creds[options.address] && this.block_to_creds[options.address].s3_params;
+                    const block_id = block_md.id;
+                    const block_dir = get_block_internal_dir(block_id);
+                    const disable_metadata = this.block_to_creds[options.address].disable_metadata;
+                    const encoded_md = disable_metadata ? '' : Buffer.from(JSON.stringify(block_md)).toString('base64');
+                    const write_params = this.block_to_creds[options.address] && {
+                        Bucket: this.block_to_creds[options.address].target_bucket,
+                        Key: `${this.block_to_creds[options.address].blocks_path}/${block_dir}/${block_id}`,
+                        Metadata: disable_metadata ? undefined : { noobaablockmd: encoded_md },
+                    };
+                    if (!s3_params || !write_params) {
+                        throw new Error('expected delegate_write_block to return write_params');
+                    }
+                    dbg.log1('got s3_params from block_store. writing using S3 sdk. s3_params =',
+                        _.omit(s3_params, 'secretAccessKey'));
+                    s3_params.httpOptions = _.omitBy({ agent: http_utils.get_unsecured_http_agent(s3_params.endpoint) },
+                        _.isUndefined);
+                    const s3 = new AWS.S3(s3_params);
+                    write_params.Body = data;
+                    await s3.putObject(write_params).promise();
+                    const data_length = data.length;
+                    usage = data_length ? {
+                        size: (block_md.is_preallocated ? 0 : data_length) + encoded_md.length,
+                        count: block_md.is_preallocated ? 0 : 1
+                    } : { size: 0, count: 0 };
+                    this._update_usage_stats(rpc_client, usage, options, 'WRITE');
                 } catch (err) {
-                    dbg.error('encountered error on _delegate_write_block_s3:', err);
-                    const error = _.pick(err, 'message', 'code');
-                    return rpc_client.block_store.handle_delegator_error({ error, usage, op_type: 'WRITE' }, options);
+                    dbg.error('BlockStoreS3 operation failed:', err, _.omit(this.block_to_creds[options.address], 'access_keys'));
```

Comment:
on failures, and specifically `AccessDenied`, try to get the credentials again and retry

Reply:
for both AccessDenied and STORAGE_NOT_EXIST did a LRUCache invalidate_key